onnxruntime/core/providers/webnn/builders/impl/gemm_op_builder.cc

```diff
@@ -42,13 +42,54 @@ Status GemmOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const N
     if (!GetShape(*input_defs[a_idx], a_shape, logger)) {
       return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Can not get shape of A.");
     }
+    std::vector<int64_t> b_shape;
+    if (!GetShape(*input_defs[b_idx], b_shape, logger)) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Can not get shape of B.");
+    }
+    // If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions.
+    bool extended_a_shape = false;
+    if (a_shape.size() == 1) {
+      extended_a_shape = true;
+      a_shape.insert(a_shape.begin(), 1);
+      a = model_builder.GetBuilder().call<emscripten::val>("reshape", a,
+                                                           emscripten::val::array(GetVecUint32FromVecInt64(a_shape)));
+    }
+    // If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions.
+    bool extended_b_shape = false;
+    if (b_shape.size() == 1) {
+      extended_b_shape = true;
+      b_shape.push_back(1);
+      b = model_builder.GetBuilder().call<emscripten::val>("reshape", b,
+                                                           emscripten::val::array(GetVecUint32FromVecInt64(b_shape)));
+    }
     // The inputs of MatMul must be at least 3D for WebNN CPU backend. Use GEMM for 2D case.
     // TODO: Remove this workaround when it is fixed in Chromium.
     if (model_builder.GetWebnnDeviceType() == WebnnDeviceType::CPU && a_shape.size() == 2) {
       output = model_builder.GetBuilder().call<emscripten::val>("gemm", a, b);
     } else {
       output = model_builder.GetBuilder().call<emscripten::val>("matmul", a, b);
     }
+    // If the inputs are both 1D， reduce the output to a scalar.
+    if (extended_a_shape && extended_b_shape) {
+      output = model_builder.GetBuilder().call<emscripten::val>("reshape", output, emscripten::val::array());
+    }
+    // After matrix multiplication the prepended 1 is removed.
+    else if (extended_a_shape) {
+      std::vector<uint32_t> new_shape;
+      for (size_t i = 0; i < b_shape.size() - 2; i++) {
+        new_shape.push_back(narrow<uint32_t>(b_shape[i]));
+      }
+      new_shape.push_back(narrow<uint32_t>(b_shape.back()));
```

Comment:
Seems the output reshaping code would be simpler if we just kept the original input shape, rather than add a dimension and then remove the new dimension we just added? e.g. 

```diff
    std::vector<int64_t> a_shape;
    ...
    std::vector<int64_t> b_shape;
+   std::vector<int64_t> original_shape;
    ...
    if (a_shape.size() == 1) {
      extended_a_shape = true;
+     original_shape = a_shape;
      ...
    }
    ...
    bool extended_b_shape = false;
    if (b_shape.size() == 1) {
      extended_b_shape = true;
+     original_shape = b_shape; // It's okay to overwrite it if it was a_shape before.
      ...
    }

    ...

    // If the inputs are both 1D， reduce the output to a scalar.
    if (extended_a_shape && extended_b_shape) {
      output = model_builder.GetBuilder().call<emscripten::val>("reshape", output, emscripten::val::array());
-   }
-   else if (extended_a_shape) {
-     std::vector<uint32_t> new_shape;
-     for (size_t i = 0; i < b_shape.size() - 2; i++) {
-       new_shape.push_back(narrow<uint32_t>(b_shape[i]));
-     }
-     new_shape.push_back(narrow<uint32_t>(b_shape.back()));
-     output = model_builder.GetBuilder().call<emscripten::val>("reshape", output, emscripten::val::array(new_shape));
-   }
-   // After matrix multiplication the appended 1 is removed.
-   else if (extended_b_shape) {
-     std::vector<uint32_t> new_shape;
-     for (size_t i = 0; i < a_shape.size() - 1; i++) {
-       new_shape.push_back(narrow<uint32_t>(a_shape[i]));
-     }
-     output = model_builder.GetBuilder().call<emscripten::val>("reshape", output, emscripten::val::array(new_shape));
+   // After matrix multiplication, restore the output to the original input shape.
+   } else if (extended_a_shape || extended_b_shape) {
+     output = model_builder.GetBuilder().call<emscripten::val>(
+         "reshape",
+         output, 
+         emscripten::val::array(GetVecUint32FromVecInt64(original_shape)));
    }
```

Reply:
The final shape is incorrect, because the removed dimension is not the one we just added. e.g. [3,4,5] * [5] will get [3,4] and [4] * [3,4,5] will get [3,5].

Reply:
👍 Gotcha.

Given `a.shape = [4], b.shape = [3,4,5]`
`a.shape = [4] -> [1,4]`
So `shape [1,4] @ shape [3,4,5] = output shape [3,1,5]`
Then `output.shape = [3,1,5] -> [3,5]`

Thanks for adding.